kolibri/core/tasks/utils.py

```diff
@@ -206,3 +212,142 @@ def db_connection():
+            self.progressbar = None
+        else:
+            # Check that we are executing inside a click context
+            # as we only want to display progress bars from the command line.
+            try:
+                click.get_current_context()
+                # Coerce to an integer for safety, as click uses Python `range` on this
+                # value, which requires an integer argument
+                # N.B. because we are only doing this in Python3, safe to just use int,
+                # as long is Py2 only
+                self.progressbar = click.progressbar(length=int(total), width=0)
+            except RuntimeError:
+                self.progressbar = None
+
+    def update_progress(self, increment, message):
+        self.progress += increment
+
+        if self.progressbar:
+            # Click only enforces integers on the total (because it is implemented assuming a length)
+            self.progressbar.update(increment)
+            if message:
+                self.progressbar.label = message
+
+
+class JobProgressMixin(object):
+    """A mixin with convenience functions for displaying
+    progress to the user, and updating progress on a job.
+
+    If ran from the command line, code here displays a progress bar to the
+    user. If ran asynchronously this mixin sends results through the Job class
+    to the main job storage database. Anyone who knows the task id for the
+    job instance can check the intermediate progress by looking at the Job's progress.
+    """
+
+    def __init__(self, *args, **kwargs):
+        self.progresstracker = None
+        self.job = get_current_job()
+        super(JobProgressMixin, self).__init__(*args, **kwargs)
+
+    def update_progress(self, increment=1, message="", extra_data=None):
+        if self.progresstracker:
+            self.progresstracker.update_progress(increment, message)
+        if self.job:
+            self.job.update_progress(
+                self.job.progress + increment, self.job.total_progress
+            )
+            if extra_data and isinstance(extra_data, dict):
+                self.job.update_job_metadata(**extra_data)
+
+    def update_job_metadata(self, **kwargs):
+        if self.job:
+            for key, value in kwargs.items():
+                self.job.extra_metadata[key] = value
+            self.job.save_meta()
+
+    def check_for_cancel(self):
+        if self.job:
+            self.job.check_for_cancel()
+
+    def start_progress(self, total=100):
+        self.progresstracker = ProgressTracker(total=total)
+        return self
+
+    def __enter__(self):
+        return self.update_progress
+
+    def __exit__(self, *args, **kwargs):
+        self.progresstracker = None
+
+    def is_cancelled(self):
+        try:
+            self.check_for_cancel()
+            return False
+        except (UserCancelledError, KeyError):
+            return True
+
+    def cancel(self):
+        return self.check_for_cancel()
```

Comment:
This is copied from the original implementation, so I haven't actually changed anything here - but you're right the method name is misleading, what it's doing is reraising a handled cancel, rather than actually initiating a cancel.

Reply:
Yeap sir then we should cancel the job there.

Reply:
I think it's actually better to rename the method, as it is never used to initiate a cancel from within the task, only to reraise the cancellation once the task has finished its cleanup tasks.